rust/windows-client/src-tauri/src/gui.rs

```diff
@@ -0,0 +1,210 @@
+                            win.hide().unwrap();
+                        } else {
+                            win.show().unwrap();
+                        }
+                    }
+                    "/quit" => app.exit(0),
+                    id => {
+                        if let Some(addr) = id.strip_prefix("/resource/") {
+                            println!("TODO copy {addr} to clipboard");
+                        }
+                    }
+                }
+            }
+        })
+        .setup(|app| {
+            if let Some(_app_link) = app_link {
+                // TODO: Handle app links that we catch at startup here
+            }
+
+            app.listen_global("scheme-request-received", |event| {
+                // TODO: Handle "firezone://handle_client_auth_callback/?client_csrf_token=bogus_csrf_token&actor_name=Bogus+Name&client_auth_token=bogus_auth_token"
+                println!("got scheme request {:?}", event.payload());
+            });
+
+            // From https://github.com/FabianLars/tauri-plugin-deep-link/blob/main/example/main.rs
+            let handle = app.handle();
+            tauri_plugin_deep_link::register("firezone", move |request| {
+                dbg!(&request);
+                handle.trigger_global("scheme-request-received", Some(request));
+            })?;
+            Ok(())
+        })
+        .build(tauri::generate_context!())?
+        .run(|_app_handle, event| {
+            if let tauri::RunEvent::ExitRequested { api, .. } = event {
+                // Don't exit if we close our main window
+                // https://tauri.app/v1/guides/features/system-tray/#preventing-the-app-from-closing
+
+                api.prevent_exit();
+            }
+        });
+    Ok(())
+}
+
+#[tauri::command]
+async fn apply_account_id(account_id: &str) -> StdResult<(), String> {
+    // These functions are wrapped since `anyhow::Error` doesn't implement `serde::Serialize` and so it can't go back to JS
+    apply_account_id_inner(account_id)
+        .await
+        .map_err(|e| format!("{e}"))
+}
+
+#[tauri::command]
+async fn apply_advanced_settings(
+    auth_base_url: &str,
+    api_url: &str,
+    log_filter: &str,
+) -> StdResult<(), String> {
+    apply_advanced_settings_inner(auth_base_url, api_url, log_filter)
+        .await
+        .map_err(|e| format!("{e}"))
+}
+
+async fn apply_account_id_inner(account_id: &str) -> Result<()> {
+    let dirs = crate::cli::get_project_dirs()?;
+    let dir = dirs.config_local_dir();
+    tokio::fs::create_dir_all(dir).await?;
+    let path = dir.join("account_id.json");
+
+    let j = serde_json::json!({
+        "account_id": account_id,
+    });
+
+    // TODO: Do something more robust. This will corrupt the file if we lose power while writing. Not the end of the world, but annoying.
```

Comment:
Sorry if this is long, it's one of my favorite topics.

Also, I don't see the Tauri store plugin having any kind of protection against this. Not even the most basic `rename` call - https://github.com/tauri-apps/plugins-workspace/blob/v1/plugins/store/src/store.rs#L206-L207
I wonder if they are aware of this or not.

Yes, I found some info on writing files safely here:
1. https://stackoverflow.com/questions/7054844/is-rename-atomic
2. https://unix.stackexchange.com/questions/297632/is-it-broken-to-replace-an-existing-file-without-fsync

I directly observed file write problems on a previous project which was for ext4 / Linux, but I assume similar things can happen to NTFS / Windows:

1. If we were appending to a log file and lost power, the log file would often have NUL bytes at the end. This seemed to happen more on older kernels but might have happened on any kernel.
2. If we unpacked a directory from a tarball and renamed the directory into a location, then pulled the power, the directory might be in its final location but empty.

Based on these observations, this is how I understand file systems / disks:
1. The FS is a distributed system like any other, if you don't put barriers (fsync), there's no "happens before" guarantees at all. You can see a write return and then a rename return, and after power loss the rename can complete but the write can be rolled back, and this is legal because it makes the FS faster in general cases.
2. Appending data to a file involves 2 or 3 steps (Allocate blocks, update metadata, and write data) which probably have no ordering guarantees between them, and the FS is only required to protect its own integrity if it loses power, not the integrity of user data inside files.
3. It's best to assume that a power loss _will_ cause the hard disk to travel back in time at least a few seconds, maybe a whole minute, and that anything that isn't carefully synced will have happened in the wrong order
4. The SQLite team already learned all this years ago and that's why it's embedded in every phone and why they advertise it as "an alternative to fopen()". SQLite already calls fsync for you and the default commit behavior is to fsync all data before returning to the user.

Without SQLite, I think the full cycle for writing an atomic file is supposed to be:
1. Delete the temp file if it already exists
2. Write the file to a temp location
5. fsync so that the temp file fully reaches disk
6. Rename the file into place (rename is typically atomic on Linux FSes, but I don't know what guarantees Windows can make about atomicity - If Windows silently performs a copy, that could be bad.)
7. fsync again so that this change is flushed to disk before performing any other I/O operations that happen to logically depend on this one having completed.

Worst of all, some hardware vendors just lie about fsync and if that primitive is broken, you can't really work around it in software:
- Postgres had an email thread about broken fsync in 2018 https://danluu.com/fsyncgate/
- The SQLite team mention that some hard disks and many USB drives completely lie about their sync and you can't even guarantee atomicity in software at that point, much less durability: https://sqlite.org/howtocorrupt.html (3.1)